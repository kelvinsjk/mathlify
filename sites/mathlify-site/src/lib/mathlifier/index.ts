/**
 * generates djot string,
 * using JavaScript's template literal syntax to handle mathematical expressions (both static and dynamic)
 *
 * math: starts with ${}, terminates with \n,
 * display: starts with $${}, terminates with \n\n,
 * display env: starts with $${'align'}, etc.
 * text: starts with @${}, terminates immediately
 */
export function mathlify(strings: TemplateStringsArray, ...values: unknown[]): string {
	return mathlifyGen(modules)(strings, ...values);
}

const modules: Modules = {
	math: (x: string) => `$\`${x}\``,
	display,
	mathEnvs: {
		equation: (x: string) => display(insertEnv('equation', x)),
		align: (x: string) => display(insertEnv('align', x)),
		gather: (x: string) => display(insertEnv('gather', x)),
		alignat: (x: string) => display(insertEnv('alignat', x)),
		equationStar: (x: string) => display(insertEnv('equation*', x)),
		alignStar: (x: string) => display(insertEnv('align*', x)),
		gatherStar: (x: string) => display(insertEnv('gather*', x)),
		alignatStar: (x: string) => display(insertEnv('alignat*', x)),
	},
};

function insertEnv(env: string, content: string, args = ''): string {
	if (env === 'alignat' || env === 'alignat*') {
		let firstLine = content.split('\n')[0];
		if (!firstLine) {
			firstLine = content.split('\n')[1];
		}
		const twoN = firstLine.split('&').length;
		if (twoN % 2 !== 0)
			console.warn(`alignat should have an even number of columns, but got ${twoN} instead`);
		args = `{${Math.floor(twoN / 2)}}`;
	}
	// todo: handle this with regex
	while (content.endsWith('\n') || content.endsWith('\t') || content.endsWith(' ')) {
		content = content.slice(0, content.length - 1);
	}
	return `\\begin{${env}}${args}${content}\n\\end{${env}}`;
}

function display(x: string): string {
	return `$$\`${x}\``;
}

enum Modes {
	text, // 0
	math, // 1
	display, // 2
	mathEnv, // 3
}

export interface Modules {
	math: (x: string) => string;
	display: (x: string) => string;
	mathEnvs?: {
		equation?: (x: string) => string;
		align?: (x: string) => string;
		gather?: (x: string) => string;
		alignat?: (x: string) => string;
		equationStar?: (x: string) => string;
		alignStar?: (x: string) => string;
		gatherStar?: (x: string) => string;
		alignatStar?: (x: string) => string;
	};
}

enum MathEnvs {
	equation,
	equationStar,
	align,
	alignStar,
	gather,
	gatherStar,
	alignat,
	alignatStar,
}

const mathEnvsKeys = [
	'equation',
	'equation*',
	'align',
	'align*',
	'gather',
	'gather*',
	'alignat',
	'alignat*',
] as const;
function isMathEnv(x: string): x is (typeof mathEnvsKeys)[number] {
	return (mathEnvsKeys as readonly string[]).includes(x);
}

const envToMode: Record<(typeof mathEnvsKeys)[number], MathEnvs> = {
	equation: MathEnvs.equation,
	'equation*': MathEnvs.equationStar,
	align: MathEnvs.align,
	'align*': MathEnvs.alignStar,
	gather: MathEnvs.gather,
	'gather*': MathEnvs.gatherStar,
	alignat: MathEnvs.alignat, // triggered by $${'alignat'}{x}
	'alignat*': MathEnvs.alignatStar, // triggered by $${'alignat'}{x}
};

interface MathEnvOptions {
	mathEnv: MathEnvs;
	eqnColumns?: number;
}

// TODO: alignat cols

export function mathlifyGen(
	modules: Modules,
): (strings: TemplateStringsArray, ...values: unknown[]) => string {
	return (strings: TemplateStringsArray, ...values: unknown[]) => {
		let currentModeString = '';
		let output = '';
		let mode = Modes.text;
		let mathEnvOptions: MathEnvOptions = { mathEnv: MathEnvs.equation };
		strings.forEach((str, i) => {
			const nextVal = values[i] ?? '';
			if (mode === Modes.text) {
				if (str.endsWith('@')) {
					// continues text mode
					currentModeString += `${str.slice(0, str.length - 1)}${nextVal}`;
				} else {
					// starts new text environment
					[mode, output, currentModeString, mathEnvOptions] = startNewEnv(
						str,
						'',
						nextVal,
						currentModeString,
					);
				}
			} else if (mode === Modes.math) {
				// checks for \n or \r\n non-greedily
				const regex = /([^]*?)(\r?\n)([^]*)/;
				const match = str.match(regex);
				if (match) {
					// end math mode
					const [, before, newline, after] = match;
					output += `${before}`;
					if (output) {
						currentModeString += modules.math(output);
					}
					[mode, output, currentModeString, mathEnvOptions] = startNewEnv(
						after,
						newline,
						nextVal,
						currentModeString,
					);
				} else {
					// continue math mode
					if (str.endsWith(`@`)) {
						// TODO: check this
						output += `${str.slice(0, str.length - 1)}`;
					} else {
						output += `${str}${nextVal}`;
					}
					if (i === strings.length - 1) {
						// final string
						currentModeString += modules.math(output);
					}
				}
			} else if (mode === Modes.display) {
				// checks for \n\n or \r\n\r\n within str. group into (#1)(#2) where #2 = \n or \r\n and it is done non-greedily
				const regex = /([^]*?)(\r?\n[ \t]*\r?\n)([^]*)/;
				const match = str.match(regex);
				if (match) {
					// end display mode
					const [, before, newline, after] = match;
					output += before;
					if (output) {
						currentModeString += modules.display(output);
					}
					[mode, output, currentModeString, mathEnvOptions] = startNewEnv(
						after,
						newline,
						nextVal,
						currentModeString,
					);
				} else {
					// continue display mode
					if (str.endsWith(`@`)) {
						// TODO: check this
						output += `${str.slice(0, str.length - 1)}`;
					} else {
						output += `${str}${nextVal}`;
					}
					if (i === strings.length - 1) {
						// final string
						currentModeString += modules.display(output);
					}
				}
			} else if (mode === Modes.mathEnv) {
				// checks for \n\n or \r\n\r\n within str. group into (#1)(#2) where #2 = \n or \r\n and it is done non-greedily
				const regex = /([^]*?)(\r?\n[ \t]*\r?\n)([^]*)/;
				const match = str.match(regex);
				if (match) {
					// end mathEnv mode
					const [, before, newline, after] = match;
					output += before;
					if (output) {
						currentModeString += insertMathEnv(mathEnvOptions, output, modules);
					}
					[mode, output, currentModeString, mathEnvOptions] = startNewEnv(
						after,
						newline,
						nextVal,
						currentModeString,
					);
				} else {
					// continue mathEnv mode
					if (str.endsWith(`@`)) {
						// TODO: check this.
						output += `${str.slice(0, str.length - 1)}`;
					} else {
						output += `${str}${nextVal}`;
					}
					if (i === strings.length - 1) {
						// final string
						currentModeString += insertMathEnv(mathEnvOptions, output, modules);
					}
				}
			}
		});
		return currentModeString;
	};
}

// returns [Mode, acc, curr, {mathEnv, eqnColumns?}]
function startNewEnv(
	after: string,
	newline: string,
	nextVal: unknown,
	curr: string,
): [Modes, string, string, MathEnvOptions] {
	const defaultMathEnvOptions = { mathEnv: MathEnvs.equation };
	let mathEnvOptions: MathEnvOptions = defaultMathEnvOptions;
	if (after.endsWith('$')) {
		let x = `${nextVal}`;
		let mode = Modes.display;
		if (isMathEnv(x)) {
			// new math env
			mode = Modes.mathEnv;
			mathEnvOptions = { mathEnv: envToMode[x] };
			//TODO: handle alignat cols
			x = '';
		}
		// new display mode
		return [mode, x, curr + `${newline}${after.slice(0, after.length - 1)}`, mathEnvOptions];
	} else if (after.endsWith('@')) {
		// new text mode
		return [
			Modes.text,
			'',
			curr + `${newline}${after.slice(0, after.length - 1)}${nextVal}`,
			defaultMathEnvOptions,
		];
	} else {
		// new math mode
		return [Modes.math, `${nextVal}`, curr + newline + after, defaultMathEnvOptions];
	}
}

function insertMathEnv(mathEnvOptions: MathEnvOptions, acc: string, modules: Modules): string {
	const mathEnv = mathEnvOptions.mathEnv;
	if (mathEnv === MathEnvs.equation) {
		return modules.mathEnvs?.equation ? modules.mathEnvs.equation(acc) : acc;
	} else if (mathEnv === MathEnvs.equationStar) {
		return modules.mathEnvs?.equationStar ? modules.mathEnvs.equationStar(acc) : acc;
	} else if (mathEnv === MathEnvs.align) {
		return modules.mathEnvs?.align ? modules.mathEnvs.align(acc) : acc;
	} else if (mathEnv === MathEnvs.alignStar) {
		return modules.mathEnvs?.alignStar ? modules.mathEnvs.alignStar(acc) : acc;
	} else if (mathEnv === MathEnvs.gather) {
		return modules.mathEnvs?.gather ? modules.mathEnvs.gather(acc) : acc;
	} else if (mathEnv === MathEnvs.gatherStar) {
		return modules.mathEnvs?.gatherStar ? modules.mathEnvs.gatherStar(acc) : acc;
	} else if (mathEnv === MathEnvs.alignat) {
		return modules.mathEnvs?.alignat ? modules.mathEnvs.alignat(acc) : acc;
	} else if (mathEnv === MathEnvs.alignatStar) {
		return modules.mathEnvs?.alignatStar ? modules.mathEnvs.alignatStar(acc) : acc;
	}
	console.warn(`mathEnv ${mathEnv} not found`);
	return acc;
}
