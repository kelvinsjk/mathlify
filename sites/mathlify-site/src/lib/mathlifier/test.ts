import { mathlify } from './';

const str = mathlify`This is inline math: ${'x+2'}xyz${'k'}
and this is text.

This is display math: $${'x+2'}xyz
abcd
${'k'}

This is aligned:
$${'align'}
  x+2 &= ${'k'} \\\\
  y &= ${'z'}`;

console.log(str);
