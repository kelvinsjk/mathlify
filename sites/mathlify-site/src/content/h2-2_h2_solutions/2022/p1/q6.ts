import { Answer } from '$lib/classes/answer';
import { mathlifierDj as mathlifier } from 'mathlifier';
import { Polynomial, quotient } from 'mathlify';
import { fractionalInverseWorking } from '$content/h2-1_h2_learn/01_functions/02_inverse/03_formula/02_practice';
import { selfInverseWorking } from '$content/h2-1_h2_learn/01_functions/04_more/01_self-inverse/02_practice';
import { QED } from '$lib/typesetting/utils';

export const answer = new Answer();

const num = new Polynomial(['a', 'k']);
const den = new Polynomial([1, [-1, 'a']]);
const f = quotient(num, den);

// a
{
	const { quotient, remainder, result } = num.longDivide(den);
	const soln = mathlifier`$${{}} ${f} = ${result}

- Translate the curve ${{}} {y=\\frac{1}{x}}
by ${'a'}
units in the positive ${'x\\text{-axis}'}
direction
- Scale the resulting curve by a factor of ${remainder}
parallel to the ${'y\\text{-axis}'}
- Translate the resulting curve by ${quotient}
units in the positive ${'y\\text{-axis}'}
direction`;
	const ans = mathlifier`Translate the curve ${{}} {y=\\frac{1}{x}}
by ${'a'}
units in the positive ${'x\\text{-axis}'}
direction.
\\
Scale the resulting curve by a factor of ${remainder}
parallel to the ${'y\\text{-axis}.'}
\\
Translate the resulting curve by ${quotient}
units in the positive ${'y\\text{-axis}'}
direction.`;
	answer.addPart(ans, soln);
}
// b
{
	const { working, exp: fInv } = fractionalInverseWorking(f, false, {
		reportInverse: true,
		swapNum: true,
		qed: true
	});
	const ans = mathlifier`${{}} f^{-1}(x) = ${fInv}.`;
	answer.addPart(ans, working);
}
// c
{
	const soln = mathlifier`We observe from part (b) that ${{}} {f(x) = f^{-1}(x).}
Hence $${'align*'} f^2(x) &= ff(x)
\\\\ &= ff^{-1}(x)
\\\\ &= x \\; ${QED}`;
	const ans = mathlifier`${{}}f^2(x)=x.`;
	answer.addPart(ans, soln);
}
// d
{
	const { ans, soln } = selfInverseWorking(f, 2023, { x: 1, qed: true });
	answer.addPart(ans, soln);
}
