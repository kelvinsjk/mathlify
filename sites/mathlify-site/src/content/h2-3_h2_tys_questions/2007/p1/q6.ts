/* The question here is the copyright of 
- Singapore Examinations and Assessment Board
- Cambridge Assessment International Education
- Ministry of Education, Singapore
- Government of Singapore

We provide this question on Mathlify under fair use for archival purposes.
These are questions more than 10 years old and are difficult to obtain
commercially.

For the newer questions, please purchase a copy of the
"ten year series" from authorized publishers and distributors.
*/

import { Question } from '$lib/classes/question';
import { mathlifierDj as mathlifier } from 'mathlifier';
import { Vector } from 'mathlify/vectors';

export const question = new Question();

// a: i-j+2k
// b: 2i+4j+k
// c: -4i+2j+2k
const a = new Vector(1, -1, 2, { stringMode: 'ijk' });
const b = new Vector(2, 4, 1, { stringMode: 'ijk' });
const c = new Vector(-4, 2, 2, { stringMode: 'ijk' });

question.addBody(
	mathlifier`Referred to the origin ${'O'},
the position vectors of the points ${'A'}
and ${'B'}
are

$${{}} ${a} \\quad \\text{and} \\quad ${b}

respectively.`
);
// a
{
	question.addPart(
		mathlifier`Show that the ${'OA'}
is perpendicular to ${'OB'}.`,
		{ marks: 2 }
	);
}
// b
{
	question.addPart(
		mathlifier`Find the position vector of the
point ${'M'}
on the line segment ${'AB'}
such that ${{}} {AM:MB = 1: 2.}`,
		{ marks: 3 }
	);
}
// c
{
	question.addPart(
		mathlifier`The position vector ${'C'}
has position vector ${{}} {${c}.}
Use a vector product to find the exact
area of triangle ${'OAC.'}`,
		{ marks: 4 }
	);
}
