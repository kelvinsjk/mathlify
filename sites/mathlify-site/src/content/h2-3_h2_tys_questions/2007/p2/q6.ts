/* The question here is the copyright of 
- Singapore Examinations and Assessment Board
- Cambridge Assessment International Education
- Ministry of Education, Singapore
- Government of Singapore

We provide this question on Mathlify under fair use for archival purposes.
These are questions more than 10 years old and are difficult to obtain
commercially.

For the newer questions, please purchase a copy of the
"ten year series" from authorized publishers and distributors.
*/

import { Question } from '$lib/classes/question';
import { modified } from '$lib/utils/typesetting/utils';
import { mathlifierDj as mathlifier } from 'mathlifier';

export const question = new Question();
const pA = 24;
const pB = 0.3;

question.addBody(
	mathlifier`@${modified}
	
	In a large population, ${pA}\\%
have a particular gene ${'A'},
and ${pB.toFixed(1)}\\%
have gene ${'B'}`
);
// a
question.addPart(
	mathlifier`Find the probability that, in a random
sample of ${10}
people from the population, at most ${4}
have gene ${'A'}.`,
	{ marks: 2 }
);
// removed approximation from old syllabus
// b (new : nested binomial)
question.addPart(
	mathlifier`\\* ${100}
samples of ${10}
people are taken from
the population. Find the probability that more than ${90}
of these samples have at most ${4}
having gene ${'A'}.`,
	{ marks: 3 }
);
// c (adapted from (ii)
question.addPart(
	mathlifier`\\* A random sample of ${1000}
people is taken from the population.
Find the probability that at least ${2}
but fewer than ${5}
have gene ${'B'}.`,
	{ marks: 2 }
);
