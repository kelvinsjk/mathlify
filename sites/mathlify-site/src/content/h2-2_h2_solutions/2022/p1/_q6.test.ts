import { expect, it } from 'vitest';
import { answer } from './q6';
import { mathlifierDj as mathlifier } from 'mathlifier';

const ans = answer.answer;
const actual: typeof ans = {
	parts: [
		{
			body: mathlifier`Translate the curve ${''}{y=\\frac{1}{x}}
by ${'a'}
units in the positive ${{}} x\\text{-axis}
direction.
\\
Scale the resulting curve by a factor of ${{}} a^2 + k
parallel to the ${{}} y\\text{-axis}.
\\
Translate the resulting curve by ${'a'}
units in the positive ${{}} y\\text{-axis}
direction.`
		},
		{
			// TODO: change xa to ax
			body: mathlifier`${{}} f^{-1}(x) = \\frac{xa+k}{x-a}.`
		},
		{
			body: mathlifier`${{}} f^2(x) = x.`
		},
		{
			body: mathlifier`${{}} f^{2023}(1) = \\frac{a+k}{1-a}.`
		}
	]
};

it('2022p1q6', () => {
	expect(ans).toEqual(actual);
	expect(answer).toMatchSnapshot();
});
