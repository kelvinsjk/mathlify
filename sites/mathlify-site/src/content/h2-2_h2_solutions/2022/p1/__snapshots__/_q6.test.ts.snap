// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`2022p1q6 1`] = `
Answer {
  "answer": {
    "parts": [
      {
        "body": "Translate the curve $\`  {y=\\frac{1}{x}} \`
by $\` a \`
units in the positive $\` x\\text{-axis} \`
direction.
\\
Scale the resulting curve by a factor of $\` a^2 + k \`
parallel to the $\` y\\text{-axis}. \`
\\
Translate the resulting curve by $\` a \`
units in the positive $\` y\\text{-axis} \`
direction.",
      },
      {
        "body": "$\`  f^{-1}(x) = \\frac{xa + k}{x - a}. \`",
      },
      {
        "body": "$\` f^2(x)=x. \`",
      },
      {
        "body": "$\` f^{2023}(1) = \\frac{a + k}{1 - a}. \`",
      },
    ],
  },
  "solution": {
    "parts": [
      {
        "body": "$$\`  \\frac{ax + k}{x - a} = a + \\frac{a^2 + k}{x - a} \`

- Translate the curve $\`  {y=\\frac{1}{x}} \`
by $\` a \`
units in the positive $\` x\\text{-axis} \`
direction
- Scale the resulting curve by a factor of $\` a^2 + k \`
parallel to the $\` y\\text{-axis} \`
- Translate the resulting curve by $\` a \`
units in the positive $\` y\\text{-axis} \`
direction",
      },
      {
        "body": "$$\` \\begin{gather*} \\text{Let }  y = \\frac{ax + k}{x - a}
	\\\\ y\\left( x - a \\right) = ax + k
	\\\\ yx - ya = ax + k
	\\\\ yx - ax = ya + k
	\\\\ x\\left( y - a \\right) = ya + k
	\\\\ x = \\frac{ya + k}{y - a} \\\\ f^{-1}(x) = \\frac{xa + k}{x - a} \\; \\blacksquare
\\end{gather*} \`",
      },
      {
        "body": "We observe from part (b) that $\`  {f(x) = f^{-1}(x).} \`
Hence $$\` \\begin{align*} f^2(x) &= ff(x)
\\\\ &= ff^{-1}(x)
\\\\ &= x \\; \\QED
\\end{align*} \`",
      },
      {
        "body": "$$\` \\begin{align*} f^{2023}(1) &= f^{2021}ff(1)
\\\\ &= f^{2021}ff^{-1}(1)
\\\\ &= f^{2021}(1)
\\\\ &= \\dotsb
\\\\ &= f(1) \\\\ &= \\frac{a + k}{1 - a} \\; \\QED
\\end{align*} \`",
      },
    ],
  },
}
`;
